multisig-wallet/project/multisig-contract/src/utils.sw

```diff
@@ -0,0 +1,105 @@
+/// Returns the b256 value of the recovered address.
+pub fn recover_signer(message_hash: b256, signature_info: SignatureInfo) -> b256 {
+    let formatted_message = match signature_info.message_format {
+        MessageFormat::None => message_hash,
+        MessageFormat::EIP191PersonalSign => eip_191_personal_sign_format(message_hash),
+    };
+
+    let prefixed_message = match signature_info.message_prefix {
+        MessagePrefix::None => formatted_message,
+        MessagePrefix::Ethereum => ethereum_prefix(formatted_message),
+    };
+
+    match signature_info.wallet_type {
+        WalletType::Fuel => {
+            let recover_result = ec_recover_address(signature_info.signature, prefixed_message);
+            require(recover_result.is_ok(), "ec_recover_address failed");
+            recover_result.unwrap().value
+        },
+        WalletType::EVM => {
+            let recover_result = ec_recover_evm_address(signature_info.signature, prefixed_message);
+            require(recover_result.is_ok(), "ec_recover_evm_address failed");
+            recover_result.unwrap().value
+        },
+    }
+}
+
+/// Creates an EIP-191 compliant transaction hash, of the version:
+/// 0x45, personal sign.
+/// It takes a data_to_sign to represent the <data to sign> in the following EIP-191 format:
+/// 0x19 <1 byte version> <version specific data> <data to sign>
+fn eip_191_personal_sign_format(data_to_sign: b256) -> b256 {
+    let initial_byte = 0x19u8;
+    let version_byte = 0x45u8;
+
+    let encoded_data = encode_data(initial_byte, version_byte, data_to_sign);
+    let encoded_data = (
+        encoded_data.get(0).unwrap(),
+        encoded_data.get(1).unwrap(),
+        encoded_data.get(2).unwrap(),
+        encoded_data.get(3).unwrap(),
+        encoded_data.get(4).unwrap(),
+    );
+
+    // Keccak256 hash the first 34 bytes of encoded_data
+    let mut result_buffer = b256::min();
+    asm(hash: result_buffer, ptr: encoded_data, bytes: 34) {
+        k256 hash ptr bytes;
+        hash: b256
+    }
+}
+
+/// Encode the packed_bytes, version_byte and message_hash into a Vec<u64> of length 40 bytes,
+/// where the first 34 bytes are the desired data tightly packed.
+fn encode_data(initial_byte: u64, version_byte: u64, message_hash: b256) -> Vec<u64> {
+    let mut data = Vec::with_capacity(5);
+
+    let (message_1, message_2, message_3, message_4) = decompose(message_hash);
+
+    data.push((initial_byte << 56) + (version_byte << 48) + (message_1 >> 16));
+
+    data.push((message_1 << 48) + (message_2 >> 16));
+    data.push((message_2 << 48) + (message_3 >> 16));
+    data.push((message_3 << 48) + (message_4 >> 16));
+    data.push(message_4 << 48);
+
+    data
+}
+
+/// Get a tuple of 4 u64 values from a single b256 value.
+fn decompose(val: b256) -> (u64, u64, u64, u64) {
+    asm(r1: __addr_of(val)) { r1: (u64, u64, u64, u64) }
+}
+
+/// Applies the prefix used by Geth to a message hash.
```

Comment:
No; EIP-191 isn't and nor is the prefix, but the prefix originates from Geth as I understand it.

Reply:
Presumably there are other ethereum clients that implement this signature scheme though? I think the prefixes are a standard across implementations. It may be that the client usually attaches the prefix but it's not necessarily the case. One could prepare a signature without using any client software if they follow the spec.

Reply:
Yh; this scheme is common across clients and wallets, but can be followed manually with the spec. I meant that as the EIP-191 document itself states:

Additionally, 0x19 has been chosen because since ethereum/go-ethereum#2940 , the following is prepended before hashing in personal_sign:
`"\x19Ethereum Signed Message:\n" + len(message)`

It seemed that it's use, with regards to EIP-191, stems from it's integration into Geth